caput/config.py

```diff
@@ -397,6 +403,81 @@ def _prop(val):
     return prop
 
 
+def logging_config(default={}):
+    """
+    A Property type that validates the caput logging config.
+
+    Allows the type to be either a string (for backward compatibility) or a dict setting log levels
+    per module.
+
+    Parameters
+    ----------
+    default : optional
+        The optional default value.
+
+    Returns
+    -------
+    prop : Property
+        A property instance setup to validate the type.
+
+    Examples
+    --------
+    Should be used like::
+
+        class Project(object):
+
+            loglevels = logging_config({"root": "INFO", "annoying.module": "WARNING"})
+    """
+
+    def _prop(val):
+        config = val
+        if isinstance(config, str):
+            config = {"root": config}
+        elif not isinstance(config, dict):
+            raise ValueError(
+                "Expected a string or YAML block for config value 'logging', got "
+                "'{}'.".format(type(config.__name__))
+            )
+
+        # check entries, get module names and warn for duplicates when sorting into new dict
+        checked_config = {}
+        for key, level in config.items():
+            level = level.upper()
+            loglevels = ["DEBUG", "INFO", "WARNING", "ERROR", "NOTSET"]
+            if level not in loglevels:
+                raise ValueError(
+                    "Expected one of {} for log level of {} (was {}).".format(
+                        loglevels, key, level
+                    )
+                )
+            if key == "root":
+                checked_config["root"] = level
+            else:
+                module = misc.import_module(key)
+                if module is None:
+                    raise ValueError(
+                        "Module referenced by logging config key {} not found.".format(
+                            key
+                        )
+                    )
+
+                already_set_to = checked_config.get(module, None)
+                if already_set_to is not None and already_set_to != level:
+                    logger.warning(
+                        "Setting log level for {} to {}, but is already set to "
+                        "{} for module {}. The old value will get ignored.".format(
+                            key, level, already_set_to, module
+                        )
+                    )
+
+                checked_config[module] = level
```

Comment:
I would recommend removing this check. There's no requirement that logger names have to map to fully qualified Python names (modules, classes etc), and definitely no requirement that they map to modules (although I now see that your function works for both). For one thing I could see us using this mechanism to turn off logging from a specific class somewhere.

Also, although it is certainly not best practice sometimes loading modules can have side effects and so I'm also not sure that importing them in here in a good idea. It would be unexpected at the very least.

Anyway, I'd still recommend removing this check. Although it works within the recommended way of using Python logging, it could break perfectly allowable logging patterns.

Reply:
meh. I was torn between the 2 options:
- Help people find their typos in the logging config
- Not having to do all this.

I decided for this because it probably gets imported anyways later and because the version number thing does something similar. But I'll remove it